tests/components/bsblan/test_config_flow.py

```diff
@@ -112,3 +113,319 @@ async def test_user_device_exists_abort(
+            CONF_PASSWORD: "admin1234",
+        },
+    )
+
+    assert result2.get("type") is FlowResultType.CREATE_ENTRY
+    assert result2.get("title") == format_mac("00:80:41:19:69:90")
+    assert result2.get("data") == {
+        CONF_HOST: "10.0.2.60",
+        CONF_PORT: 80,
+        CONF_PASSKEY: "1234",
+        CONF_USERNAME: "admin",
+        CONF_PASSWORD: "admin1234",
+    }
+    assert "result" in result2
+    assert result2["result"].unique_id == format_mac("00:80:41:19:69:90")
+
+    assert len(mock_setup_entry.mock_calls) == 1
+    assert len(mock_bsblan.device.mock_calls) == 1
+
+
+async def test_zeroconf_discovery_no_mac_in_announcement(
+    hass: HomeAssistant,
+    mock_bsblan: MagicMock,
+    zeroconf_discovery_info_no_mac: Mock,
+) -> None:
+    """Test Zeroconf discovery works when no MAC address is in the announcement."""
+    # Create a mock object that mimics ZeroconfServiceInfo but allows properties_raw
+    discovery_info = zeroconf_discovery_info_no_mac
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_ZEROCONF},
+        data=discovery_info,
+    )
+
+    assert result["type"] == FlowResultType.FORM
+    assert result["step_id"] == "discovery_confirm"
+
+    # Now complete the discovery by providing credentials and connecting
+    result2 = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        {
+            CONF_USERNAME: "admin",
+            CONF_PASSWORD: "secret",
+        },
+    )
+
+    await hass.async_block_till_done()
+
+    assert result2["type"] == FlowResultType.CREATE_ENTRY
+    assert result2["title"] == "00:80:41:19:69:90"  # MAC from fixture file
+    assert result2["data"] == {
+        CONF_HOST: "10.0.2.60",
+        CONF_PORT: 80,
+        CONF_PASSKEY: None,
+        CONF_USERNAME: "admin",
+        CONF_PASSWORD: "secret",
+    }
+
+
+async def test_zeroconf_discovery_connection_error(
+    hass: HomeAssistant,
+    mock_bsblan: MagicMock,
+    zeroconf_discovery_info: ZeroconfServiceInfo,
+) -> None:
+    """Test connection error during zeroconf discovery shows the correct form."""
+    mock_bsblan.device.side_effect = BSBLANConnectionError
+
+    discovery_info = zeroconf_discovery_info
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN,
+        context={"source": SOURCE_ZEROCONF},
+        data=discovery_info,
+    )
+
+    assert result.get("type") is FlowResultType.FORM
+    assert result.get("step_id") == "discovery_confirm"
+
+    result2 = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={
+            CONF_PASSKEY: "1234",
+            CONF_USERNAME: "admin",
+            CONF_PASSWORD: "admin1234",
+        },
+    )
+
+    assert result2.get("type") is FlowResultType.FORM
+    assert result2.get("step_id") == "discovery_confirm"
+    assert result2.get("errors") == {"base": "cannot_connect"}
```

Comment:
Let's finish this one, so we know its able to recover

Reply:
I’m not quite sure what you’re asking.

Reply:
I added some code to try again. Is this what you wanted?